ENH: support for indices of type long added

jor- opened the pull request:
A support for indices of type `long` is added. This allows to handle big matrices (more than 2^31 entries or dimension larger then 2^31). The needed index type is determined automatically by default or can be passed explicitly.

(Furthermore `Factor._print` is added analogous to `Common._print`)

jor- commented:
Since some tests are still faulty, maybe some other can give me a hint?

I basically replaced the `cholmod_` calls with `cholmod_l_` calls when `long` is used.

I am not quite sure about the relation to the `itype` constant in the `cholmod common` object. Must the `cholmod_` functions be used, when `itype` is `int` and `cholmod_l_` when `itype` is `long` or are these independent of each other?

anntzer commented:
I'm a bit worried that the tests fail even though you have not changed them :-)  At least you should add some simple tests for both cases.
A quick reading of the docs suggests that your interpretation is correct (set `itype` to `int` whe using `cholmod`, to `long` when using `cholmod_l`)... but I don't know much about this.

jor- commented:
I managed to correct the errors.

The problem was that for example in
```
if self._use_long:
    cholmod_start = cholmod_l_start
cholmod_start(&self._common)
```
``cholmod_l_start`` was called even though ``self._use_long`` was ``False``. I do not know if this was some kind of bug, but it worked using a new function name like:
```
if self._use_long:
    cholmod_c_start = cholmod_l_start
else
    cholmod_c_start = cholmod_start
cholmod_c_start(&self._common)
```

jor- commented:
I have added some tests for indices of type ``long``.

anntzer commented on sksparse/cholmod.pyx:
```diff
@@ -86,18 +95,29 @@ cdef extern from "cholmod.h":
         size_t d
         void * x
-        int xtype, dtype
+        int xtype
```
Leave them on the same line?  Actually also merge the size_t declarations above, and the int fields declared for the Common class.

jor- replied:
Okay, I merged some lines in the ``struct`` definitions. I moved also all ``struct`` before the function definitions to make the code clearer.

anntzer commented on sksparse/cholmod.pyx:
```diff
@@ -233,7 +310,12 @@ cdef class _DenseCleanup:
+            cholmod_c_free_dense = cholmod_l_free_dense
+        else:
+            cholmod_c_free_dense = cholmod_free_dense
+
```
remove this empty line

jor- replied:
Done.

anntzer commented on sksparse/cholmod.pyx:
```diff
@@ -453,17 +601,31 @@ cdef class Factor:
         assert C_perm
         try:
-            cholmod_updown(not subtract, C_perm, self._factor,
+            cholmod_c_updown(not subtract, C_perm, self._factor,
```
align the line below; similarly for line 661, 679, etc.

jor- replied:
Done.

anntzer commented on sksparse/cholmod.pyx:
```diff
@@ -764,10 +947,10 @@ cdef class Factor:
                                format="csc"))
 
-def analyze(A, mode="auto", ordering_method="default"):
+def analyze(A, mode="auto", ordering_method="default", use_long=None):
```
Do we really need to expose `use_long`?  Isn't it enough to just rely on the input type?

jor- replied:
The problem is that the input type is not the only criterion. It is possible, that the input matrix ``A`` uses 32-bit indices which might be reasonable if the number of non-zero entries and the shape is in the 32-bit integer range. But it might be that the cholesky factor ``L`` has much more non-zero entries so that 64-bit indices are needed here. This is not known apriori. If the user known that 32-bit indices are not enough, he/she can force the use of 64-bit indices with ``use_long=True``.

anntzer replied:
Ah, I see.

btw, you have a test failure.

anntzer commented on sksparse/cholmod.pyx:
```diff
@@ -265,42 +347,84 @@ cdef void _error_handler(
+        else:
+            cholmod_c_start = cholmod_start
+        cholmod_c_start(&self._common)
+        assert (_use_long == 0 and self._common.itype == CHOLMOD_INT) or (_use_long == 1 and self._common.itype == CHOLMOD_LONG)
```
Let's keep the lines short.

jor- replied:
Done.

anntzer commented:
You have a test failure now.

jor- commented:
Yes, I will take a look at the failure.

I removed the ``test_integer_size`` which tests 32-bit integers to be the default index type in ``scipy`` since 64-bit integers are also supported now with this PR.

jor- commented:
The ``D`` matrix is not correct for the test decomposition with ``mode="supernodal"`` and ``int64`` indices. With ``mode="simplicial"`` and ``int64`` indices it works. For me it looks more like a ``cholmod`` bug. Do you have an idea about this?

The ``L L*`` decomposition does work with both ``modes`` and indice types.

jor- commented:
I think I found the error.

The problem is in the ``D`` function:
```
cdef int * super_ = <int *> self._factor.super_
cdef int * pi = <int *> self._factor.pi
cdef int * px = <int *> self._factor.px
```
must be 
```
cdef long * super_ = <long *> self._factor.super_
cdef long * pi = <long *> self._factor.pi
cdef long * px = <long *> self._factor.px
```
if  64-bit indices are used.

But something like 
```
if self._common._use_long:
    cdef long * super_ = <long *> self._factor.super_
    cdef long * pi = <long *> self._factor.pi
    cdef long * px = <long *> self._factor.px
else:
    cdef int * super_ = <int *> self._factor.super_
    cdef int * pi = <int *> self._factor.pi
    cdef int * px = <int *> self._factor.px
```
is not allowed in ``cython``. Do you have an idea how to do this?

anntzer commented:
I think you need to delegate to an additional (private, cdef only) method that takes a fusedtype as argument.

The issue you discovered also calls for carefully checking whether the same issue affects any other methods...

jor- commented:
I rebased this branch to the updated master branch and fixed the issue with ``int`` and ``long`` in the ``D`` function. Now all tests are successful. I also checked the entire code: I have not seen an other issue where ``int`` is always uses instead of ``int`` resp.  ``long`` depending on the index type.

jor- commented:
Do you have further 	comments? Should I squash the commits?

anntzer commented on sksparse/cholmod.pyx:
```diff
@@ -42,7 +42,7 @@ np.import_array()
 cdef extern from "numpy/arrayobject.h":
     void PyArray_ENABLEFLAGS(np.ndarray arr, int flags)
-
+    
```
blank spaaaaaces :)

jor- replied:
removed

anntzer commented:
Did you manage to build scikit-sparse on Windows?  I gave it a try some time ago and couldn't figure out how to do it.  I'd be very interested in a solution.

Please squash the commits, yes (otherwise I'll do it myself and you won't get to pick the commit message :-))

jor- commented:
I did not try it on Windows, because I have no Windows installed.

It is squashed.

jor- commented:
Since the last pull requests ``scikit-sparse`` got some significant improvements. Are you going to release a new version on PyPI (https://pypi.python.org/pypi/scikits.sparse)?

anntzer commented:
Can you make a separate PR that updates `doc/changes.rst` for both of your PRs (under a 0.4.0 section)?  I'll do a release after that.

You can also smuggle in the following whitespace changes :-)
```
diff --git a/setup.py b/setup.py
index 5eec968..75e63b0 100644
--- a/setup.py
+++ b/setup.py
@@ -43,13 +43,13 @@ if __name__ == "__main__":
           url = URL,
           long_description = LONG_DESCRIPTION,
           classifiers =
-            [ 'Development Status :: 3 - Alpha',
-              'Environment :: Console',
-              'Intended Audience :: Developers',
-              'Intended Audience :: Science/Research',
-              'License :: OSI Approved :: GNU General Public License (GPL)',
-              'Topic :: Scientific/Engineering'],
-          setup_requires=['setuptools>=18.0', 'numpy', 'cython'],
+            ['Development Status :: 3 - Alpha',
+             'Environment :: Console',
+             'Intended Audience :: Developers',
+             'Intended Audience :: Science/Research',
+             'License :: OSI Approved :: GNU General Public License (GPL)',
+             'Topic :: Scientific/Engineering'],
+          setup_requires = ['setuptools>=18.0', 'numpy', 'cython'],
           # You may specify the directory where CHOLMOD is installed using the
           # library_dirs and include_dirs keywords in the lines below.
           ext_modules = [Extension("sksparse.cholmod", ["sksparse/cholmod.pyx"],
diff --git a/sksparse/test_cholmod.py b/sksparse/test_cholmod.py
index 408634c..9306777 100644
--- a/sksparse/test_cholmod.py
+++ b/sksparse/test_cholmod.py
@@ -85,7 +85,6 @@ def convert_matrix_indices_to_long_indices(matrix):
     matrix.indptr = np.asarray(matrix.indptr, dtype=np.int64)
     return matrix
 
-
 def test_complex():
     c = complex_matrix()
     fc = cholesky(c)
```

jor- commented:
Okay, I am going to make a separate PR for version 0.4 and include your whitespace changes.

I have an additional question: Is the ``scikit-sparse`` project under ``2-clause BSD`` or ``GPL``  license? ``2-clause BSD`` is mentioned in ``README`` and ``GPL`` in ``setup.py``. This is somehow inconsistent. I can change it to one licence with the PR for version 0.4.

I would also like to add  myself to the copyright notice since I made several major contributions. Is this okay for you?

anntzer commented:
See https://github.com/scikit-sparse/scikit-sparse#license for the license situation.

I'd rather just replace all the mentions of copyright by `Copyright (C) 2008-2016 The scikit-sparse developers` and add a section after `Contact` in the docs with a chronological list of all committers.
